elements/lisk-genesis/src/validate.ts

```diff
@@ -0,0 +1,329 @@
+} from './utils';
+import {
+	EMPTY_BUFFER,
+	GENESIS_BLOCK_GENERATOR_PUBLIC_KEY,
+	GENESIS_BLOCK_MAX_BALANCE,
+	GENESIS_BLOCK_REWARD,
+	GENESIS_BLOCK_SIGNATURE,
+	GENESIS_BLOCK_TRANSACTION_ROOT,
+} from './constants';
+import { getHeaderAssetSchemaWithAccountAsset } from './utils/schema';
+
+export const validateGenesisBlock = (
+	block:
+		| GenesisBlock
+		| {
+				// eslint-disable-next-line @typescript-eslint/no-explicit-any
+				[key: string]: any;
+		  },
+	options: { roundLength: number; accountAssetSchema?: Schema },
+): ErrorObject[] => {
+	const { header, payload } = block as GenesisBlock;
+
+	// Genesis block schema validation, only check payload length to be zero
+	const payloadErrors = validator.validate(genesisBlockSchema, {
+		header: EMPTY_BUFFER,
+		payload,
+	});
+
+	// Genesis block header validation
+	const headerErrors = [
+		...validator.validate(genesisBlockHeaderSchema, {
+			...header,
+			asset: EMPTY_BUFFER,
+		}),
+	];
+
+	// Custom header validation not possible with validator
+	if (header.generatorPublicKey !== GENESIS_BLOCK_GENERATOR_PUBLIC_KEY) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.generatorPublicKey',
+			schemaPath: 'properties.generatorPublicKey',
+			params: { allowedValue: GENESIS_BLOCK_GENERATOR_PUBLIC_KEY },
+		});
+	}
+
+	if (header.reward !== GENESIS_BLOCK_REWARD) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.reward',
+			schemaPath: 'properties.reward',
+			params: { allowedValue: GENESIS_BLOCK_REWARD },
+		});
+	}
+
+	if (header.signature !== GENESIS_BLOCK_SIGNATURE) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.signature',
+			schemaPath: 'properties.signature',
+			params: { allowedValue: GENESIS_BLOCK_SIGNATURE },
+		});
+	}
+
+	if (header.transactionRoot !== GENESIS_BLOCK_TRANSACTION_ROOT) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.transactionRoot',
+			schemaPath: 'properties.transactionRoot',
+			params: { allowedValue: GENESIS_BLOCK_TRANSACTION_ROOT },
+		});
+	}
+
+	// Genesis block asset validation
+	const assetSchemaWithAccountAsset = getHeaderAssetSchemaWithAccountAsset(
+		genesisBlockHeaderAssetSchema,
+		options.accountAssetSchema ?? defaultAccountAssetSchema,
+	);
+	const assetErrors = [
+		...validator.validate(assetSchemaWithAccountAsset, header.asset),
+	];
+
+	const initDelegates = [...header.asset.initDelegates];
+	const accountAddresses = [];
+	const delegateAddresses = [];
+	let totalBalance = BigInt(0);
+
+	for (const account of header.asset.accounts) {
+		accountAddresses.push(account.address);
+		totalBalance += BigInt(account.balance);
+
+		if (account.asset.delegate.username !== '') {
+			delegateAddresses.push(account.address);
+		}
+
+		if (
+			account.publicKey !== undefined &&
```

Comment:
If it's correct account structure, public key will never be undefined. so it should be caught by the above validation.